### ISSUE TITLE
TransferManager UploadFile crashes

### ISSUE THREAD

**Issue by drivehappy** [aws/aws-sdk-cpp#665]
We have some code that uses UploadFile with a shared_ptr<IOStream>, built from an in-memory buffer, which looks like this:

```
void myFunc()
  String ss(...);
  shared_ptr<StringStream> awsSS {new StringStream()};
	awsSS->str(ss);
	shared_ptr<IOStream> awsIOS {awsSS};
	auto transferHandle = m_awsContext->m_transferMgr->UploadFile(
		awsIOS,
		...);
	return make_shared<AwsFileUpload>(bucketKeyName, m_bucketConfig.prefix, transferHandle);
}
```

After updating from an older build 1.0.31 to 1.1.53 this is crashing due to an Access Violation / segfault with what appears to be related to how the awsIOS data internally is now handled. This seems to be related to refactoring changes in 580f160 that removes the reference count on the shared_ptr stream being passed in. In our case that goes out of scope before the worker has a chance to complete, and then it reads invalid memory.

This crashes both under Windows 10 x64 (VS 2017) and Ubuntu 17.04 (GCC 6.3).

I verified that at least release 1.1.9 (Jul 11) did not exhibit this problem, while 1.1.11 did (which included this change).

I attempted to fix this myself in a fork, however it seems that the aws-cpp-sdk-transfer-tests are failing on my system (GCC 6.3). I couldn't find any information on whether these tests expect real infrastructure for testing, or how to set that up (e.g. is it expected that these tests are failing?). Would it be sufficient to back out the above refactoring change in a PR without a regression test?

**Comment by marcomagdy** [comment]
This is the same issue reported in #651. We're currently working on a fix for this. It will patched soon.

**Comment by drivehappy** [comment]
Thanks, I had noticed that issue but wasn't entirely sure if it was the same - I'll go ahead and close this one.

**Comment by drivehappy** [comment]
Re-opening after re-reading the linked issue I don't think it's quite the same. This one is calling out a specific issue with the `Aws:IOStream` lifetime in the `TransferManager::UploadFile` call. The API is providing a `shared_ptr<Aws::IOStream>` to be passed in, indicating that the AWS library will handle lifetime of the `Aws::IOStream` correctly, however the underlying implementation is performing:
```
return this->DoUploadFile(fileStream.get(), ...);
```
Which loses the reference count through the `.get()`. This means the user of the library must manage the lifetime of said object, which seems contrary to what the API is stating with the use of `shared_ptr<Aws::IOStream>`.

Additionally, it looks like there were some changes for #647 around some of the callbacks (those changes do not solve this issue).

**Comment by drivehappy** [comment]
I put together a few changes in a branch that fix this for me: https://github.com/drivehappy/aws-sdk-cpp/commit/43f6b7aa23216e6b4be7d9e84a89bd62f460ea89

Again, I'm concerned about a PR at the moment as I can't get the transfer tests to run from master on my machine (unsure if real buckets are needed to get them to pass). I would like to add a regression test for this if possible.

**Comment by marcomagdy** [comment]
You're right. We'll get it fixed. Thank you

**Comment by marcomagdy** [comment]
Fixed in [v1.2.7](https://github.com/aws/aws-sdk-cpp/releases/tag/1.2.7)
